### PR TITLE
Fix the problem that hifi4 does not support XT_ROUND_S

### PR DESCRIPTION
**Pull request by hongshui3000**
fix https://github.com/thesofproject/sof/issues/4042

Signed-off-by: yxh yangxiaohuamail@gmail.com

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test this please** to run this test once.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by dbaluta** [comment]
@hongshui3000 thanks a lot for taking care of this. 

Few style comments:
- commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
- please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003

The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?

**Comment by hongshui3000** [comment]
> @hongshui3000 thanks a lot for taking care of this.
> 
> Few style comments:
> 
> * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> 
> The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?

The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
This is my understanding of this problem

**Comment by iuliana-prodan** [comment]
> > @hongshui3000 thanks a lot for taking care of this.
> > Few style comments:
> > 
> > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > 
> > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> 
> The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> This is my understanding of this problem

If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
From my understanding this should run on all xtensa architectures.

**Comment by hongshui3000** [comment]
> > > @hongshui3000 thanks a lot for taking care of this.
> > > Few style comments:
> > > 
> > > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > > 
> > > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> > 
> > 
> > The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> > Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> > This is my understanding of this problem
> 
> If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
> From my understanding this should run on all xtensa architectures.

XT_ROUND_S is only for xtensa processors, not for processors of other architectures, so pcm_converter_generic is not changed. pcm_converter_generic is written in general C code and can be compiled theoretically for any architecture. My changes are only for the xtensa architecture

**Comment by hongshui3000** [comment]
@dbaluta 
checkpatch：
```
ERROR: Unrecognized email address: 'yxh yangxiaohuamail@gmail.com'
#12: 
Signed-off-by: yxh yangxiaohuamail@gmail.com
```
Is this my error, or is it another problem?

**Comment by dbaluta** [comment]
@hongshui3000 signed-off-by line should look like this:

`Signed-off-by: John Doe <john.doe@gmail.com>`

**Comment by hongshui3000** [comment]
> @hongshui3000 signed-off-by line should be formatted like this:
> 
> `Signed-off-by: John Doe <john.doe@gmail.com>`

understood

**Comment by hongshui3000** [comment]
@dbaluta 
checkpatch：
```
WARNING: 'doesn'' may be misspelled - perhaps 'doesn't'?
#70: FILE: src/audio/pcm_converter/pcm_converter_hifi3.c:706:
+	/* add 0.5 to round correctly but assert it doesn't lead to overflow */

total: 0 errors, 1 warnings, 141 lines checked
```
This doesn't seem to be my problem.
????

**Comment by paulstelian97** [comment]
> @dbaluta
> checkpatch：
> 
> ```
> WARNING: 'doesn'' may be misspelled - perhaps 'doesn't'?
> #70: FILE: src/audio/pcm_converter/pcm_converter_hifi3.c:706:
> +	/* add 0.5 to round correctly but assert it doesn't lead to overflow */
> 
> total: 0 errors, 1 warnings, 141 lines checked
> ```
> 
> This doesn't seem to be my problem.
> ????

Does sound like a false positive to me.

**Comment by dbaluta** [comment]
@hongshui3000 it is a false positive.

**Comment by iuliana-prodan** [comment]
> > > > @hongshui3000 thanks a lot for taking care of this.
> > > > Few style comments:
> > > > 
> > > > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > > > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > > > 
> > > > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> > > 
> > > 
> > > The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> > > Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> > > This is my understanding of this problem
> > 
> > 
> > If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
> > From my understanding this should run on all xtensa architectures.
> 
> XT_ROUND_S is only for xtensa processors, not for processors of other architectures, so pcm_converter_generic is not changed. pcm_converter_generic is written in general C code and can be compiled theoretically for any architecture. My changes are only for the xtensa architecture

Your change is only for HIFI3, but the commit message says HIFI4. There's a mismatch here.

**Comment by hongshui3000** [comment]
> > > > > @hongshui3000 thanks a lot for taking care of this.
> > > > > Few style comments:
> > > > > 
> > > > > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > > > > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > > > > 
> > > > > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> > > > 
> > > > 
> > > > The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> > > > Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> > > > This is my understanding of this problem
> > > 
> > > 
> > > If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
> > > From my understanding this should run on all xtensa architectures.
> > 
> > 
> > XT_ROUND_S is only for xtensa processors, not for processors of other architectures, so pcm_converter_generic is not changed. pcm_converter_generic is written in general C code and can be compiled theoretically for any architecture. My changes are only for the xtensa architecture
> 
> Your change is only for HIFI3, but the commit message says HIFI4. There's a mismatch here.

This file is indeed hifi3, but it can also support hifi4. All my tests are done by hifi4. That’s why I pointed out my platform hifi4 in the submission information

**Comment by paulstelian97** [comment]
@iuliana-prodan Do you remember how to compile with our 2020 toolchain to see if these patches fix build for our platforms? It would be interesting to check.

**Comment by iuliana-prodan** [comment]
> > > > > > @hongshui3000 thanks a lot for taking care of this.
> > > > > > Few style comments:
> > > > > > 
> > > > > > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > > > > > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > > > > > 
> > > > > > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> > > > > 
> > > > > 
> > > > > The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> > > > > Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> > > > > This is my understanding of this problem
> > > > 
> > > > 
> > > > If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
> > > > From my understanding this should run on all xtensa architectures.
> > > 
> > > 
> > > XT_ROUND_S is only for xtensa processors, not for processors of other architectures, so pcm_converter_generic is not changed. pcm_converter_generic is written in general C code and can be compiled theoretically for any architecture. My changes are only for the xtensa architecture
> > 
> > 
> > Your change is only for HIFI3, but the commit message says HIFI4. There's a mismatch here.
> 
> This file is indeed hifi3, but it can also support hifi4. All my tests are done by hifi4. That’s why I pointed out my platform hifi4 in the submission information
> I am not sure that I need to write a separate c code for hifi4
> I always suspect that this problem will be encountered with hifi3, but I don't use the hifi3 platform, so I can't verify my idea.

@hongshui3000 What platform with hifi4 are you using?

**Comment by hongshui3000** [comment]
> > > > > > > @hongshui3000 thanks a lot for taking care of this.
> > > > > > > Few style comments:
> > > > > > > 
> > > > > > > * commit message should follow some rules: Please read https://chris.beams.io/posts/git-commit/#seven-rules
> > > > > > > * please fix checkpatch.pl errors: https://github.com/thesofproject/sof/pull/4057/checks?check_run_id=2386830003
> > > > > > > 
> > > > > > > The thing I want to understand is if do we really want to make the code written to run on HIFI3 on HIFI4?
> > > > > > 
> > > > > > 
> > > > > > The problem is not hifi4 or hifi3, the problem is that the xtensa architecture is a configurable architecture, the code should make some reasonable judgments based on the configuration, otherwise it is difficult to be compatible with all xtensa configurations.
> > > > > > Therefore, the entire coding process needs to pay attention to the commonalities and characteristics between the architectures
> > > > > > This is my understanding of this problem
> > > > > 
> > > > > 
> > > > > If the problem is not HIFI3 or HIFI 4, why didn't you updated also the [pcm_converter_generic](https://github.com/thesofproject/sof/blob/main/src/audio/pcm_converter/pcm_converter_generic.c)?
> > > > > From my understanding this should run on all xtensa architectures.
> > > > 
> > > > 
> > > > XT_ROUND_S is only for xtensa processors, not for processors of other architectures, so pcm_converter_generic is not changed. pcm_converter_generic is written in general C code and can be compiled theoretically for any architecture. My changes are only for the xtensa architecture
> > > 
> > > 
> > > Your change is only for HIFI3, but the commit message says HIFI4. There's a mismatch here.
> > 
> > 
> > This file is indeed hifi3, but it can also support hifi4. All my tests are done by hifi4. That’s why I pointed out my platform hifi4 in the submission information
> > I am not sure that I need to write a separate c code for hifi4
> > I always suspect that this problem will be encountered with hifi3, but I don't use the hifi3 platform, so I can't verify my idea.
> 
> @hongshui3000 What platform with hifi4 are you using?

I use a hifi4 platform that my company configures (hardware), which may be different from NXP and INTEL

**Comment by dbaluta** [comment]
@hongshui3000 care to share with us your company name. At least in private :). We also have an IRC channel. #sof on irc.freenode.net (you can use a webclient like this: https://webchat.freenode.net/) 

The reason we are asking this is that we are interested to see if there are other people looking at SOF and to share our future scenarios :).

**Comment by hongshui3000** [comment]
> @hongshui3000 care to share with us your company name. At least in private :). We also have an IRC channel. #sof on irc.freenode.net (you can use a webclient like this: https://webchat.freenode.net/)
> 
> The reason we are asking this is that we are interested to see if there are other people looking at SOF and to share our future scenarios :).

I enter freenode. you are?

**Comment by lgirdwood** [comment]
@zrombel everything OK with CI ? Looks like scheduler has failed but the internal CI has passed ??

**Comment by lgirdwood** [comment]
@hongshui3000 I've sent you an invite to join SOF developer group. This mean CI will run your PRs. Please check email for invite.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
> > > This file is indeed hifi3, but it can also support hifi4. All my tests are done by hifi4. That’s why I pointed out my platform hifi4 in the submission information
> > > I am not sure that I need to write a separate c code for hifi4
> > > I always suspect that this problem will be encountered with hifi3, but I don't use the hifi3 platform, so I can't verify my idea.
> > 
> > 
> > @hongshui3000 What platform with hifi4 are you using?
> 
> I use a hifi4 platform that my company configures (hardware), which may be different from NXP and INTEL
> xtensa is a dsp that can be highly customized.
> Generally speaking, my solution is crude and not elegant. I also hope someone has a better solution

I've reviewed the general C code and it looks fine, however I'm no expert with Hifi intrinsics so need @singalsu to approve before we merge.

One question, @singalsu do we need a 
```c
#if (CONFIG_HIFI4)
```

**Review by cujomalainey — APPROVED** [review]
LGTM, only request to go the extra mile would be a couple unit tests

**Comment by lgirdwood** [comment]
> LGTM, only request to go the extra mile would be a couple unit tests

And UTs can now be run on host with local GCC - see ```./scripts/run-mocks.sh```

**Comment by zrombel** [comment]
Scheduler shouldn't be reported here and it can be ignored :) Build and codecheck are ok. PR is good to merge.

File: src/audio/pcm_converter/pcm_converter_hifi3.c [hunk 1]
```diff
@@ -658,12 +658,67 @@ static void pcm_convert_s16_to_f_lin(const void *psrc, void *pdst,
...
+{
+	int64_t dd = d;
+
+	if (a > 32)
+		a = 32;
+	else if (a < -32)
+		a = -32;
+
+	dd = a >= 0 ? dd << a : dd >> -a;
+	if (dd > INT32_MAX)
```
Comment (singalsu) [inline src/audio/pcm_converter/pcm_converter_hifi3.c:676]: Argument d is signed type so should you also check for INT32_MIN? If it's always non-negative, please add a comment about that.

File: src/audio/pcm_converter/pcm_converter_hifi3.c [hunk 2]
```diff
@@ -658,12 +658,67 @@ static void pcm_convert_s16_to_f_lin(const void *psrc, void *pdst,
...
+ *            operations library inclusion by compiler, when in whole topology
+ *            only external component needs float input.
+ * \param pow additional exponent component,
+ *            number of fractional bits in fixed point value.
+ *            Use '0' for normal conversion to integers
+ * \return (int32_t)src * 2**pow
+ */
+static int32_t _pcm_convert_f_to_i(int32_t src, int32_t pow)
+{
+	int32_t exponent, mantissa, dst;
```
Comment (singalsu) [inline src/audio/pcm_converter/pcm_converter_hifi3.c:699]: Code style, minor, we usually declare one variable at time.

File: src/audio/pcm_converter/pcm_converter_hifi3.c [hunk 3]
```diff
@@ -658,12 +658,67 @@ static void pcm_convert_s16_to_f_lin(const void *psrc, void *pdst,
...
+
+	exponent = (src >> 23);
+	exponent = (exponent & 0xFF) + pow - 127; /* exponential */
+	mantissa = BIT(23) | (MASK(22, 0) & src); /* mantisa + 1.0 [Q9.22] */
```
Comment (singalsu) [inline src/audio/pcm_converter/pcm_converter_hifi3.c:703]: Typo in comment?

Comment (hongshui3000) [reply src/audio/pcm_converter/pcm_converter_hifi3.c:703]: Will correct

File: src/audio/pcm_converter/pcm_converter_hifi3.c (hunk 2 above)
Comment (hongshui3000) [reply src/audio/pcm_converter/pcm_converter_hifi3.c:699]: Will correct

File: src/audio/pcm_converter/pcm_converter_hifi3.c (hunk 1 above)
Comment (hongshui3000) [reply src/audio/pcm_converter/pcm_converter_hifi3.c:676]: function ：_pcm_convert_f_to_i	This is called here, so it is a positive number . I will add a comment
```
static int32_t _pcm_convert_f_to_i(int32_t src, int32_t pow)
{
	int32_t exponent;
	int32_t mantissa;
	int32_t dst;

	exponent = (src >> 23);
	exponent = (exponent & 0xFF) + pow - 127; /* exponential */
	mantissa = BIT(23) | (MASK(22, 0) & src); /* mantissa + 1.0 [Q9.22] */
	/* calculate power */
	dst = _pcm_shift(mantissa, exponent - 23);
	/* add 0.5 to round correctly but assert it doesn't lead to overflow */
	if (exponent - 22 < 9 || (src & BIT(31)) == BIT(31))
		dst += _pcm_shift(mantissa, exponent - 22) & 1;
	/* copy sign to dst */
	dst = (dst ^ (src >> 31)) + (int)((unsigned int)src >> 31);

	return dst;
}
```

**Comment by hongshui3000** [comment]
https://github.com/thesofproject/sof/pull/4079

**Comment by dbaluta** [comment]
@hongshui3000 indeed #4079  looks better. Can you close this PR?

**Comment by lgirdwood** [comment]
ok, I'll close this now to avoid confusion with #4079 as the current preferred PR. This can always be reopened if needed.

**Pull request closed without merging**